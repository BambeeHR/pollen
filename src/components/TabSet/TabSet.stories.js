import CounterBadge from '../CounterBadge/CounterBadge.vue';
import TabSet from './TabSet.vue';

export default {
  title: 'TabSet',
  component: TabSet,
};

export const Default = () => ({
  components: { TabSet },
  data() {
    return {
      options: [
        {
          label: 'Foo',
          content: 'This is the simplest way to display content.',
        },
        {
          label: 'Bar',
          content:
            'Pass Tabset an array of objects with a `label` and `content` key.',
        },
      ],
    };
  },
  template: `<TabSet :options="options" />`,
});

export const WithContentAsDynamicSlots = () => ({
  components: { TabSet },
  data() {
    return {
      options: [
        {
          label: 'Foo',
          value: 'foo',
        },
        {
          label: 'Bar',
          value: 'bar',
        },
      ],
    };
  },
  template: `
  <div>
    <TabSet :options="options">
      <template v-for="{ label, value } in options" v-slot:[value]>
        <div :key="value" class="bg-gray-5 p-8">Hello from {{ label }}</div>
      </template>
    </Tabset>
    <p class="mt-4">
      If you want to customize the TabPanel, dynamically named slots are
      provided. The slot name will match the value key of each option.
    </p>
  </div>
  `,
});

export const WithContentAsSlots = () => ({
  components: { TabSet },
  data() {
    return {
      options: ['Foo', 'Bar', 'Hello World'],
    };
  },
  template: `
  <div>
    <TabSet :options="options">
      <template v-slot:Foo>
        <div class="bg-gray-5 p-8">Hello from Foo</div>
      </template>
      <template v-slot:Bar>
        <div class="bg-primary p-8 text-white">Hello from Foo</div>
      </template>
      <template v-slot:['Hello World']>
        <div class="bg-success p-8 text-white">Hello from Hello world</div>
      </template>
    </Tabset>
    <p class="mt-4">
      If you want to customize the TabPanel, dynamically named slots are
      provided. The slot name will match the value key of each option. You can
      also simply pass an array of strings to the option prop. Strings will be
      mapped to objects with label and value keys.
    </p>
  </div>
  `,
});

export const WithLabelsPassedAsComponents = () => {
  return {
    components: { TabSet },
    data() {
      return {
        firstCount: 2,
        secondCount: 12,
      };
    },
    computed: {
      options() {
        return [
          {
            label: {
              components: { CounterBadge },
              template: `<span class="flex items-center">Foo <CounterBadge :value="${this.firstCount}" class="ml-1" /></span>`,
            },
            value: 'foo',
            content: 'Hello from Foo',
          },
          {
            label: {
              components: { CounterBadge },
              template: `<span class="flex items-center">Bar <CounterBadge :value="${this.secondCount}" class="ml-1" /></span>`,
            },
            value: 'bar',
            content: 'Hello from Bar',
          },
        ];
      },
    },
    template: `
    <TabSet :options="options" />
    `,
  };
};
