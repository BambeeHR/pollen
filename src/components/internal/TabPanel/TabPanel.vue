<template>
  <div
    :id="id"
    class="tab-panel"
    :class="{ 'tab-panel--active': active }"
    role="tabpanel"
    :aria-hidden="!active"
    :hidden="!active"
  >
    <slot />
  </div>
</template>

<script>
import kebabCase from 'lodash/kebabCase';
import shortid from 'shortid';

export default {
  props: {
    active: {
      type: Boolean,
      default: false,
    },
    tabsetId: {
      type: String,
      default: shortid.generate,
    },
    value: {
      type: String,
      default: shortid.generate,
    },
  },
  computed: {
    id() {
      return `${this.tabsetId}-${kebabCase(this.value).toLowerCase()}-panel`;
    },
  },
};
</script>

<style scoped>
.tab-panel {
  @apply hidden;
}

.tab-panel.tab-panel--active {
  @apply block;
}
</style>
