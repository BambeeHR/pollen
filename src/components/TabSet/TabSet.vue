<template>
  <div :id="id" class="tab-set">
    <TabControls
      v-model="activeTab"
      :full-width="fullWidth"
      :tabset-id="id"
      :options="mappedOptions"
    />
    <TabPanel
      v-for="{ value, content } in mappedOptions"
      :key="value"
      :active="value === activeTab"
      :tabset-id="id"
      :value="value"
    >
      <slot :name="value" :activeTab="activeTab">
        <div class="tab-set__simple-content">{{ content }}</div>
      </slot>
    </TabPanel>
  </div>
</template>

<script>
import shortid from 'shortid';
import TabControls from '../TabControls/TabControls.vue';
import TabPanel from '../internal/TabPanel/TabPanel.vue';
import mapOptions from '../../util/mapOptions';

export default {
  components: { TabControls, TabPanel },
  props: {
    fullWidth: {
      type: Boolean,
      default: false,
    },
    id: {
      type: String,
      default: shortid.generate,
    },
    initialValue: {
      type: String,
      default: null,
    },
    options: {
      type: Array,
      default: () => [],
    },
  },
  data() {
    return {
      activeTab: this.initialValue,
    };
  },
  computed: {
    mappedOptions() {
      return mapOptions(this.options);
    },
  },
  watch: {
    activeTab(newValue) {
      this.$emit('tab-change', newValue);
    },
  },
  created() {
    if (this.activeTab || !this.mappedOptions.length) return;
    this.activeTab = this.mappedOptions[0].value;
  },
};
</script>

<style scoped>
.tab-set__simple-content {
  @apply p-4;
}
</style>
