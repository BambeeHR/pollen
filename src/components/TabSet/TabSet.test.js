import { mount } from '@vue/test-utils';
import TabPanel from '../internal/TabPanel/TabPanel.vue';
import TabSet from './TabSet.vue';

const options = [
  {
    label: 'Foo',
    content: 'This is the simplest way to display content.',
  },
  {
    label: 'Bar',
    content:
      'Pass Tabset an array of objects with a `label` and `content` key.',
  },
];

describe('TabSet', () => {
  test('renders correctly', () => {
    const wrapper = mount(TabSet, {
      propsData: {
        id: 'test',
        options,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('activates first option when no initialValue is passed', () => {
    const wrapper = mount(TabSet, {
      propsData: {
        options,
      },
    });
    expect(wrapper.find('.tab-controls__tab--active').text()).toBe('Foo');
  });

  test('activates selected option when initialValue is passed', () => {
    const wrapper = mount(TabSet, {
      propsData: {
        options,
        initialValue: 'Bar',
      },
    });
    expect(wrapper.find('.tab-controls__tab--active').text()).toBe('Bar');
  });

  test('shows correct content when switching tabs', async () => {
    const wrapper = mount(TabSet);
    wrapper.setProps({ options });
    wrapper.setData({ activeTab: 'Foo' });
    await wrapper.vm.$nextTick();
    const firstPanel = wrapper.findAllComponents(TabPanel).at(0);
    const secondPanel = wrapper.findAllComponents(TabPanel).at(1);
    expect(firstPanel.element).toBeVisible();
    expect(secondPanel.element).not.toBeVisible();

    wrapper.setData({ activeTab: 'Bar' });
    await wrapper.vm.$nextTick();
    expect(firstPanel.element).not.toBeVisible();
    expect(secondPanel.element).toBeVisible();
  });
});
