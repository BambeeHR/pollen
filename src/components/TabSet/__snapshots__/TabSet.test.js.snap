// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabSet renders correctly 1`] = `
<div
  class="tab-set"
  id="test"
>
  <div
    class="tab-controls"
  >
    <ul
      class="tab-controls__tabs"
      role="tablist"
    >
      <li
        class="tab-controls__tab tab-controls__tab--active"
      >
        <a
          aria-controls="test-foo-panel"
          aria-selected="true"
          class="tab-controls__link"
          href="#test-foo-panel"
          id="test-foo-tab"
          role="tab"
        >
          Foo
        </a>
      </li>
      <li
        class="tab-controls__tab"
      >
        <a
          aria-controls="test-bar-panel"
          class="tab-controls__link"
          href="#test-bar-panel"
          id="test-bar-tab"
          role="tab"
        >
          Bar
        </a>
      </li>
    </ul>
     
    <div
      class="tab-controls__indicator"
      role="presentation"
    />
  </div>
   
  <div
    class="tab-panel tab-panel--active"
    id="test-foo-panel"
    role="tabpanel"
  >
    <div
      class="tab-set__simple-content"
    >
      This is the simplest way to display content.
    </div>
  </div>
  <div
    aria-hidden="true"
    class="tab-panel"
    hidden="hidden"
    id="test-bar-panel"
    role="tabpanel"
  >
    <div
      class="tab-set__simple-content"
    >
      Pass Tabset an array of objects with a \`label\` and \`content\` key.
    </div>
  </div>
</div>
`;
